From: Heiko Wundram <heiko.wundram@freifunk-celle.de>
Date: Mon, Feb 2 10:00:00 CET 2016 +0100
Subject: Fix frequencies for station networks if fixed channel.

This patch fixes frequencies and HT mode in station mode in case a
configuration binding the radio to specific frequencies and/or HT
modes is set up.

diff --git a/package/network/services/hostapd/files/netifd.sh b/package/network/services/hostapd/files/netifd.sh
index 30d12e9..de3de90 100644
--- a/package/network/services/hostapd/files/netifd.sh
+++ b/package/network/services/hostapd/files/netifd.sh
@@ -495,6 +495,14 @@ wpa_supplicant_add_network() {
 		[ "$_w_driver" = "nl80211" ] ||	wpa_key_mgmt="WPA-NONE"
 	}
 
+	[[ "$_w_mode" = "sta" ]] && {
+		[ -n "$channel" ] && {
+			freq="$(get_freq "$phy" "$channel")"
+			append network_data "scan_freq=$freq" "$N$T"
+			append network_data "freq_list=$freq" "$N$T"
+		}
+	}
+
 	[[ "$_w_mode" = adhoc ]] && append network_data "$_w_modestr" "$N$T"
 
 	case "$auth_type" in
@@ -573,8 +581,28 @@ wpa_supplicant_add_network() {
 		append network_data "mcast_rate=$mc_rate" "$N$T"
 	}
 
-	local ht_str
-	[ -n "$ht" ] && append network_data "htmode=$ht" "$N$T"
+	[[ "$_w_mode" = "adhoc" ]] && {
+		[ -n "$htmode" ] && {
+			append network_data "htmode=$htmode" "$N$T"
+		}
+	}
+
+	[[ "$_w_mode" = "sta" ]] && {
+		case "$htmode" in
+			NOHT)
+				append network_data "disable_ht=1" "$N$T"
+				append network_data "disable_ht40=1" "$N$T"
+				append network_data "disable_vht=1" "$N$T"
+			;;
+			HT20)
+				append network_data "disable_ht40=1" "$N$T"
+				append network_data "disable_vht=1" "$N$T"
+			;;
+			HT40*)
+				append network_data "disable_vht=1" "$N$T"
+			;;
+		esac
+	}
 
 	cat >> "$_config" <<EOF
 network={
